torch/fx/subgraph_rewriter.py

```diff
@@ -0,0 +1,202 @@
+            if n not in match.nodes_map.values():
+                return
+            for n_ in n.uses:
+                mark_nodes_for_deletion(n_)
+            nodes_to_delete.add(n)
+            # By removing the `original_graph` node's name early, we
+            # ensure that we can insert nodes from `replacement` with
+            # the same name. (This is strictly cosmetic)
+            if n.name in original_graph._used_names:
+                original_graph._used_names.pop(n.name)
+
+        mark_nodes_for_deletion(subgraph_output)
+
+        # Map replacement graph nodes to their copy in `original_graph`
+        copied_nodes: Dict[Node, Node] = {}
+
+        # Insert the new graph
+        with original_graph.inserting_before(subgraph_output):
+            for n in replacement_graph.nodes:
+                if n.op == "output":
+                    # To ensure that we have one and only one "output"
+                    # node in our final graph, we need to change our
+                    # logic slightly if the pattern extends to the end
+                    # of the graph (i.e. if we matched the pattern
+                    # "output" node to the original graph's "output"
+                    # node)
+                    if subgraph_output.op != "output":
+                        n = n.args[0]
+                        nodes_to_delete.add(copied_nodes[n])
+                    copied_nodes[n] = original_graph.node_copy(n,
+                                                               lambda n :
+                                                               copied_nodes[n])
+                    subgraph_output.replace_all_uses_with(copied_nodes[n])
+                elif n.op == "placeholder":
+                    # Temporarily change the insertion point to the
+                    # beginning of the graph to ensure that placeholder
+                    # nodes are inserted before anything else
+                    with original_graph.inserting_before(None):
+                        copied_nodes[n] = original_graph.node_copy(n,
+                                                                   lambda n :
+                                                                   copied_nodes[n])
+                        if n.name in original_graph_inputs.keys():
+                            original_node = original_graph_inputs[n.name]
+                            nodes_to_delete.add(original_node)
+                            original_node.replace_all_uses_with(copied_nodes[n])
+                            original_node = copied_nodes[n]
+                else:
+                    copied_nodes[n] = original_graph.node_copy(n,
+                                                               lambda n :
+                                                               copied_nodes[n])
+
+        for n in list(nodes_to_delete):
+            n._remove_from_list()
+            n._erased = True
+            original_graph._len -= 1
```

Comment:
Why not `original_graph.erase_node(n)`?

Reply:
We have to run through the nodes in a very specific order to ensure that we don't hit the `len(to_erase.users) > 0` case (graph.py:256). At first, it seemed like we could just do that...

```
def erase_nodes(n: Node) -> None:
     users = n.all_input_nodes
     if n in nodes_to_delete:
          original_graph.erase_node(n)
     for user in users:
          erase_nodes(user)

original_graph_output = next(iter(reversed(original_graph.nodes)))
erase_nodes(original_graph_output)
```

...but I ran into an issue. We insert the replacement subgraph before we erase the pattern subgraph in the larger graph. Because of how we hook in the "output" node during insertion, it's possible that some of the pattern nodes get cut off from the rest of the graph. I.e. the nodes will still be in `_node_list`, but they won't be reachable through any sort of graph traversal algorithm like the one above.

After getting your feedback, I decided to conditionally delete certain nodes when we add the replacement "output" node to the graph. See new version of the PR and let me know what you think.

Some other solutions:
1. Manually erase the nodes by iterating through the nodes list (what I first submitted). Pros: fast n' easy. Cons: ignores API in a big way. I agree with you that it's not ideal.
2. Call the above `erase_nodes` method in a for-loop iterating through the entire nodes list. Keep a `seen` set to prevent errors and/or O(n^2) time complexity. Pros: fits into existing program API, cons: unnecessary use of space